File: esercizi/part-iii/ricorsione/potenzacoda.c
/* potenzacoda.c */

/* Calcola la potenza tramite funzione ricorsiva.
 * Utilizza ricorsione in coda.
 */

 /* istruzioni per la compilazione e l'esecuzione:
 $ make potenzacoda
 $ ./potenzacoda
 */

#include <stdio.h>

double power(double base, int exponent);
double power_helper(double base, int exponent, double parziale);

int main(void)
{
  double b = 3.;
  int e = 4;

  printf("%g^%d = %g\n", b, e, power(b, e));

  return 0;
}

/* power:
 *
 * suppone che exponent sia un intero maggiore
 * o uguale a 1
 */
double power(double base, int exponent)
{
  return power_helper(base, exponent, 1.);
}

double power_helper(double base, int exponent, double parziale)
{
  if (exponent == 1) return base * parziale;

  return power_helper(base, exponent-1, base*parziale);
}

File: esercizi/part-iii/python/potenzacoda.py
# potenzacoda.py

# Calcola la potenza tramite funzione ricorsiva
# Utilizza la ricorsione in coda

# power:
#
# suppone che exponent sia un intero maggiore o uguale a 1
def power_helper(base, exponent, parziale):
    if (exponent == 1):
        return base * parziale

    return power_helper(base, exponent-1, base * parziale)

def power(base, exponent):
    return power_helper(base, exponent, 1)

# main

b = 3
e = 4

print "%d^%d = %d" % (b, e, power(b, e))
